sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala

```diff
@@ -338,7 +362,93 @@ abstract class GpuToTimestamp
   override def doColumnar(lhs: GpuColumnVector, rhs: Scalar): ColumnVector = {
     val tmp = if (lhs.dataType == StringType) {
       // rhs is ignored we already parsed the format
-      lhs.getBase.asTimestampMicroseconds(strfFormat)
+
+      val DAY_MICROS = 24 * 60 * 60 * 1000000L
+      val specialDates = GpuCast.calculateSpecialDates
+          .map {
+            case (name, days) => (name, days * DAY_MICROS)
+          }
+
+      def daysScalar(name: String): Scalar = {
+        Scalar.timestampFromLong(DType.TIMESTAMP_MICROSECONDS, specialDates(name))
+      }
+
+      def daysEqual(name: String): ColumnVector = {
+        lhs.getBase.equalTo(Scalar.fromString(name))
+      }
+
+      // the cuDF `is_timestamp` function is less restrictive than Spark's behavior for UnixTime
+      // and ToUnixTime and will support parsing a subset of a string so we check the length of 
+      // the string as well which works well for fixed-length formats but if/when we want to 
+      // support variable-length formats (such as timestamps with milliseconds) then we will need
+      // to use regex instead.
+      val isTimestamp = withResource(lhs.getBase.getCharLengths) { actualLen =>
+        withResource(Scalar.fromInt(sparkFormat.length)) { expectedLen =>
+          withResource(actualLen.equalTo(expectedLen)) { lengthOk =>
+            withResource(lhs.getBase.isTimestamp(strfFormat)) { isTimestamp =>
+              isTimestamp.and(lengthOk)
+            }
+          }
+        }
+      }
+
+      // in addition to date/timestamp strings, we also need to check for special dates and null
+      // values, since anything else is invalid and should throw an error or be converted to null
+      // depending on the policy
+      withResource(isTimestamp) { isTimestamp =>
+        withResource(daysEqual(GpuCast.EPOCH)) { isEpoch =>
+          withResource(daysEqual(GpuCast.NOW)) { isNow =>
+            withResource(daysEqual(GpuCast.TODAY)) { isToday =>
+              withResource(daysEqual(GpuCast.YESTERDAY)) { isYesterday =>
+                withResource(daysEqual(GpuCast.TOMORROW)) { isTomorrow =>
+                  withResource(lhs.getBase.isNull) { isNull =>
+                    val canBeConverted = isTimestamp.or(isEpoch.or(isNow.or(isToday.or(
+                      isYesterday.or(isTomorrow.or(isNull))))))
+
+                    // throw error if legacyTimeParserPolicy is EXCEPTION
+                    if (timeParserPolicy == ExceptionTimeParserPolicy) {
```

Comment:
I think we can move this checkup and save a lot of computation in some cases.  `isTimestamp` is the only thing that is common between this check and the actual conversion. It would be nice to only do the check when necessary and then close all of the temp columns before starting on the actual conversion.

Reply:
The boolean vectors `isTimestamp`, `isEpoch`, `isNow`, etc are used later on when performing the actual conversion. However, I am wondering if we should introduce a new config option to override all of this and force the expression to run onto GPU to preserve what we were able to do before with incompatibleOps, at least until we have support for a wider range of formats.

Reply:
My real concern was about extra work and extra memory usage.